### PR TITLE
Controllers actions

### PR DESCRIPTION
**Pull request by MisterJames**

### CODE REVIEW

**Comment by MisterJames** [comment]
Just noticing that there is a lot of overlap here with the next one. I'll clean them both up and get another commit in here.

**Comment by ardalis** [comment]
Ok. I'm going to try and knock out the rest of the pre-MVC ASPNET content soon as well, and see if we have any update on the REPL for ASPNET.

**Comment by MisterJames** [comment]
@ardalis I actually wrapped up results.md and I think these two flow okay as-are. Open to comments/feedback/suggestions, but otherwise these are fairly set to go.

I do have my ASP.NET project that I'm bumping along the way, but I've got it outside of the docs branch and intend to add it at the end as the completed solution.

File: content/asp.net/getting-started/controllers-actions.md [hunk 1]
```diff
@@ -1,20 +1,122 @@
...
 
 #### Sample Files
 Download a ZIP containing this tutorial's sample files:
-- [Initial Version] - Use this as a starting point when following along with the tutorial yourself
-- [Completed Version] - Includes the completed versions of all samples
+- [Completed Version (Pending)] - Includes the completed versions of all MVC tutorials
 
-See the [Issue](https://github.com/dotnet/training-tutorials/issues/60) to claim this lesson and/or view what should be included in it.
+## Controllers and Actions in ASP.NET Core
+As we talked about in the [last tutorial](mvc.md), a _controller_ is where we put the logic needed to respond to incoming requests. Each incoming request is evaluated and mapped into a specific action on a controller by virtue of its HTTP verb, path and querystring through a process called [routing](routing.md).
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: I don't think "querystring" is a word, it should be "[query string](https://en.wikipedia.org/wiki/Query_string)".

File: content/asp.net/getting-started/controllers-actions.md [hunk 2]
```diff
@@ -1,20 +1,122 @@
...
 
-Start the lesson here.
+Actions are responsible for ensuring that incoming requests are valid and that appropriate responses are returned to the client, typically building up and sending a model to a [view](views.md) for rendering. You can put whatever logic is required in an action to respond accordingly to requests, but the purpose of controllers and actions is to help separate view rendering from business logic, not to provide a place to put your business logic. Business logic should be located in services that are passed into your controller using [dependency injection](controller-dependencies.md). 
 
-...
+Each action has a role in constructing a proper HTTP response to the client, but you don't have to go it alone; the Framework has a number of features that will help you do so. You can start to  explore those features by creating a new class in the controllers folder called `PersonController`. Update the class so that it appears as below:
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: > but you don't have to go it alone

Should this be "you don't have to *do* it alone"?

File: content/asp.net/getting-started/controllers-actions.md [hunk 3]
```diff
@@ -1,20 +1,122 @@
...
+    }
+}
+``` 
+There's only a few lines of code, but what's there is pretty powerful. First off, we inherit `Controller`, which gives us the ability to access the context of the request, information about the current user, and a way to leverage validation.  `Controller` also gives us a suite of helper methods that we can use to help generate responses.
+
+Secondly, we use one of the most common controller helpers in the action when we make a call to `View()`, which triggers the rendering of the view given the current state of our controller.
+
+Next, we have a return type of `IActionResult` giving us a flexible way to generate all different types of HTTP responses, from the typical `200 OK` to more complex responses like `302 Permanent Redirect` and a whole host of others.  These responses can be created using controller helpers, which we'll examine in greater detail later in this lesson.
+
+Finally, we have a method named `Index` and, along with our class named `PersonController`, users are now able to request the URL `person/`. This works because of the conventions and defaults of the framework and project template through a mechanism called routing.
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: Routing is already introduced at the start of this section, so it feels clumsy to introduce it here again.

I think it would be nice to combine the two introductions of routing into one.

Or maybe just shorten this to something like:

>  This works because of routing conventions and defaults of the framework and project template.

File: content/asp.net/getting-started/controllers-actions.md [hunk 4]
```diff
@@ -1,20 +1,122 @@
...
+
+> Note that you can actually visit the `PersonController` if you run your app at this time and navigate to `~/Person` in the browser, but you will see an error because we haven't yet added a [view](views.md). You can, however, create a breakpoint in your controller on the `return View();` statement, which would allow you see execution flow through to that point.
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: > navigate to `~/Person`

Should this be just "`/Person`"? I'm not sure readers will know what is the `~` supposed to represent.

File: content/asp.net/getting-started/controllers-actions.md [hunk 5]
```diff
@@ -1,20 +1,122 @@
...
+| /               | Home       | Index   | There is nothing specified in the route that can be mapped. The defaults are used: `Home` for the controller and `Index` for the action.  |
+| /Home/Contact   | Home       | Contact | Both the controller and action are specified in the route.                                                                                |
+| /Person         | Person     | Index   | Whilethe controller is specified, the action is not; thus the default `Index` will be used.                                                |
+| /Person/Index/6 | Person     | Index   | The `Index` action on the `Person` controller will be invoked and the parameter `id`, if present, will be set to 6.                         |
+
+## Attribute routes
+When the globally-configured routes do not work for you you can override the defaults on a per-controller or per-action basis using routing attributes. For instance, you could decorate your `PersonController` with the following attribute to change the route from `/person` to `/peeps`:
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: I think there should be a command between "you" and "you".

File: content/asp.net/getting-started/controllers-actions.md [hunk 6]
```diff
@@ -1,20 +1,122 @@
...
+In addition to the actions that you write yourself, you're also privy to the host of helper methods that are built into the base controller that you inherit from. These methods help to formulate a repsonse after you've completed any validation, service access or processing required. We've already looked at one as part of the samples above, but let's have another quick look:
+
+```c#
+public IActionResult Index()
+{
+    return View();
+}
+```        
+
+The call to the helper method `View` above gives you an easy way to say, "I'm done processing here, and now I would like the user to see the page related to this action."  The framework takes over, maintains the state of any properties you've set on the controller and renders a view following the conventions we covered earlier.
```
Comment (svick) [inline content/asp.net/getting-started/controllers-actions.md]: What conventions is this referring to? Should this be a link?

File: content/asp.net/getting-started/controllers-actions.md (hunk 2 above)
Comment (MisterJames) [inline content/asp.net/getting-started/controllers-actions.md]: "Go it alone" is a common phrase here...may be a colloquialism though; I'll change it.

File: content/asp.net/getting-started/controllers-actions.md (hunk 6 above)
Comment (MisterJames) [inline content/asp.net/getting-started/controllers-actions.md]: Not a link, but I'll clarify as 

> ...and renders a view following the default naming and location conventions we covered earlier.

File: content/asp.net/getting-started/controllers-actions.md [hunk 7]
```diff
@@ -1,20 +1,122 @@
...
+
+| Route           | Controller | Action  | Reason                                                                                                                                    |
+|-----------------|------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------|
+| /               | Home       | Index   | There is nothing specified in the route that can be mapped. The defaults are used: `Home` for the controller and `Index` for the action.  |
+| /Home/Contact   | Home       | Contact | Both the controller and action are specified in the route.                                                                                |
+| /Person         | Person     | Index   | Whilethe controller is specified, the action is not; thus the default `Index` will be used.                                                |
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: whilethe -> while the

File: content/asp.net/getting-started/controllers-actions.md [hunk 8]
```diff
@@ -1,20 +1,122 @@
...
+
+```c#
+[Route("peeps")]
+public class PersonController : Controller
+{
+    // ...
+}
+```
+
+Likewise, an action can be setup to use a different route if you apply an action to it, for instance, with the above `peeps` route set on the controller, you can have `peeps/favs` as the route for the action below with the following route applied:
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: "if you apply an _attribute_ to it"

File: content/asp.net/getting-started/controllers-actions.md [hunk 9]
```diff
@@ -1,20 +1,122 @@
...
+
+        [Route("/people/home")]
+        public IActionResult Index()
+        {
+            //...
+        }
+    }
+```
+
+Routes can get tricky to debug, so use them judiciously and know how they affect incoming requests in your app.  For example, someone looking at the above controller might make assumptions about the `Index` action, given the default route confiruration on the project; however, with the attributes applied as above, a request to `/person/index` would give them a `404 Not Found`.  Read more about routes [here](routing.md).
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: confiruration -> configuration

File: content/asp.net/getting-started/controllers-actions.md [hunk 10]
```diff
@@ -1,20 +1,122 @@
...
+        {
+            //...
+        }
+    }
+```
+
+Routes can get tricky to debug, so use them judiciously and know how they affect incoming requests in your app.  For example, someone looking at the above controller might make assumptions about the `Index` action, given the default route confiruration on the project; however, with the attributes applied as above, a request to `/person/index` would give them a `404 Not Found`.  Read more about routes [here](routing.md). 
+
+## Controller Helpers and Result Objects 
+In addition to the actions that you write yourself, you're also privy to the host of helper methods that are built into the base controller that you inherit from. These methods help to formulate a repsonse after you've completed any validation, service access or processing required. We've already looked at one as part of the samples above, but let's have another quick look:
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: repsonse -> response

File: content/asp.net/getting-started/controllers-actions.md [hunk 11]
```diff
@@ -1,20 +1,122 @@
...
+```c#
+public IActionResult Index()
+{
+    return View();
+}
+```        
+
+The call to the helper method `View` above gives you an easy way to say, "I'm done processing here, and now I would like the user to see the page related to this action."  The framework takes over, maintains the state of any properties you've set on the controller and renders a view following the default naming and location conventions we covered earlier.
+
+Just like `View`, there are other helpers that will make the lifting easier on the controller.
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: "make the lifting easier" isn't a phrase I'm familiar with. I've heard "make life easier for" but even that doesn't sound right. How about "there are other helpers that will assist you in returning responses from your actions"

File: content/asp.net/getting-started/controllers-actions.md [hunk 12]
```diff
@@ -1,20 +1,122 @@
...
+
+ - **Redirection**- You can use `Redirect`, `LocalRedirect`, `RedirectToRoute` and other similar methods in your actions for the purpose of forwarding users on to other location, for example, after a login.
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: "to other locations" or "to another location"

File: content/asp.net/getting-started/controllers-actions.md [hunk 13]
```diff
@@ -1,20 +1,122 @@
...
+    return View();
+}
+```        
+
+The call to the helper method `View` above gives you an easy way to say, "I'm done processing here, and now I would like the user to see the page related to this action."  The framework takes over, maintains the state of any properties you've set on the controller and renders a view following the default naming and location conventions we covered earlier.
+
+Just like `View`, there are other helpers that will make the lifting easier on the controller.
+
+ - **Redirection**- You can use `Redirect`, `LocalRedirect`, `RedirectToRoute` and other similar methods in your actions for the purpose of forwarding users on to other location, for example, after a login.   
+ - **Standard Response Types**- When you are working at more of an API level or need more granular control over building the response, you can use pre-built classes like `OkResult` or `NotFoundResult` to convey the results of your processing.
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: The helpers for these are Ok() and NotFound(). They return the result types you mention, but this is a list of helpers, right?

File: content/asp.net/getting-started/controllers-actions.md [hunk 14]
```diff
@@ -1,20 +1,122 @@
...
+
+ - **Redirection**- You can use `Redirect`, `LocalRedirect`, `RedirectToRoute` and other similar methods in your actions for the purpose of forwarding users on to other location, for example, after a login.   
+ - **Standard Response Types**- When you are working at more of an API level or need more granular control over building the response, you can use pre-built classes like `OkResult` or `NotFoundResult` to convey the results of your processing.   
+ - **Json**- If you want to control the way your data is returned, perhaps decorating an object in order to better interoperate with another service or client, you can manually build JSON responses. 
+ - **PartialView**- When you only need to render part of a page without the overhead of a layout, typically used in AJAX scenarios where only a portion of the page is fetched from the server.   
+ - **ViewBag, ViewData**- These are containers that are backed by the same key/value pairs in a dictionary. `ViewData` uses a string for a key, whereas `ViewBag` is a `dynamic` that allows you to the dictionary as though the keys were property names.
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: Not sure these fit, since they're not helpers but properties. But you can probably leave them.

File: content/asp.net/getting-started/controllers-actions.md [hunk 15]
```diff
@@ -1,20 +1,122 @@
...
+ - **Json**- If you want to control the way your data is returned, perhaps decorating an object in order to better interoperate with another service or client, you can manually build JSON responses. 
+ - **PartialView**- When you only need to render part of a page without the overhead of a layout, typically used in AJAX scenarios where only a portion of the page is fetched from the server.   
+ - **ViewBag, ViewData**- These are containers that are backed by the same key/value pairs in a dictionary. `ViewData` uses a string for a key, whereas `ViewBag` is a `dynamic` that allows you to the dictionary as though the keys were property names.  
 
 ## Next Steps
 
-Give the reader some additional exercises/tasks they can perform to try out what they've just learned.
\ No newline at end of file
+ - You'll be getting an error if you try to get to an action that doesn't have a corresponding view. Views will be covered in the [next lesson](views.md).
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: You'll be getting -> You'll get

File: content/asp.net/getting-started/controllers-actions.md [hunk 16]
```diff
@@ -1,20 +1,122 @@
...
+ - Actions can return different types of results, especially when using the helper methods. Try using the browser developer tools (usually accessed by hitting F12) and experiment with different result types.
```
Comment (ardalis) [inline content/asp.net/getting-started/controllers-actions.md]: using... experimenting
or "Try using..._to_ experiment with..."

**Comment by ardalis** [comment]
:watch
